libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -98,68 +117,160 @@ m_nameDispenser(_context.dispenser)
+
+	auto rewriteAssignmentOrVariableDeclarationLeftHandSide = [this](
 		auto& _stmt,
-		auto const& _variables
+		auto& _lhsVars
 	) -> OptionalStatements {
 		using StatementType = decay_t<decltype(_stmt)>;
-		if (_stmt.value)
-			visit(*_stmt.value);
-		bool leftHandSideNeedsMoving = util::contains_if(_variables, [&](auto const& var) {
-			return m_memoryOffsetTracker(var.name);
-		});
-		if (!leftHandSideNeedsMoving)
-			return {};
-
-		langutil::SourceLocation loc = _stmt.location;
 
-		if (_variables.size() == 1)
+		langutil::SourceLocation  loc = _stmt.location;
+		if (_lhsVars.size() == 1)
 		{
-			optional<YulString> offset = m_memoryOffsetTracker(_variables.front().name);
-			yulAssert(offset, "");
-			return generateMemoryStore(
-				m_context.dialect,
-				loc,
-				*offset,
-				_stmt.value ? *std::move(_stmt.value) : Literal{loc, LiteralKind::Number, "0"_yulstring, {}}
-			);
+			optional<YulString> offset = m_memoryOffsetTracker(_lhsVars.front().name);
+			if (offset)
+				return generateMemoryStore(
+					m_context.dialect,
+					loc,
+					*offset,
+					_stmt.value ? *std::move(_stmt.value) : Literal{loc, LiteralKind::Number, "0"_yulstring, {}}
+				);
+			else
+				return {};
 		}
+		FunctionCall const* functionCall = get_if<FunctionCall>(_stmt.value.get());
+		yulAssert(functionCall, "");
+		auto rhsSlots = m_functionReturnVariables.at(functionCall->functionName.name) |
+			ranges::views::transform([&](auto const& _var) {
+				return m_memoryOffsetTracker(_var.name);
+			}) | ranges::to<vector<optional<YulString>>>;
+
+		// Nothing to do, if the right-hand-side remains entirely on the stack and
+		// none of the variables in the left-hand-side are moved.
+		if (
+			ranges::none_of(rhsSlots, [](optional<YulString> const& _slot) { return _slot.has_value(); }) &&
+			!util::contains_if(_lhsVars, [&](auto const& var) { return m_memoryOffsetTracker(var.name); })
+		)
+			return {};
 
-		VariableDeclaration tempDecl{loc, {}, std::move(_stmt.value)};
 		vector<Statement> memoryAssignments;
 		vector<Statement> variableAssignments;
-		for (auto& var: _variables)
+		VariableDeclaration tempDecl{loc, {}, std::move(_stmt.value)};
+
+		yulAssert(rhsSlots.size() == _lhsVars.size(), "");
+		for (auto&& [lhsVar, rhsSlot]: ranges::views::zip(_lhsVars, rhsSlots))
 		{
-			YulString tempVarName = m_nameDispenser.newName(var.name);
-			tempDecl.variables.emplace_back(TypedName{var.location, tempVarName, {}});
+			unique_ptr<Expression> rhs;
+			if (rhsSlot)
+				rhs = make_unique<Expression>(generateMemoryLoad(m_context.dialect, loc, *rhsSlot));
+			else
+			{
+				YulString tempVarName = m_nameDispenser.newName(lhsVar.name);
+				tempDecl.variables.emplace_back(TypedName{lhsVar.location, tempVarName, {}});
+				rhs = make_unique<Expression>(Identifier{loc, tempVarName});
+			}
 
-			if (optional<YulString> offset = m_memoryOffsetTracker(var.name))
+			if (optional<YulString> offset = m_memoryOffsetTracker(lhsVar.name))
```

Comment:
The offset of an lhs variable cannot overlap with the offset of a return variable of the called function because in order to calculate the memory needs for a function, we always look at all the variables in a function independent of the order of the internal calls, right?

Reply:
The ``StackToMemoryMover`` can assume that there are no overlaps in any assigned memory slots, that's the job of the ``StackLimitEvader`` when it assigns slots.
How the ``StackLimitEvader`` does it hasn't change - it first looks at leaves of the call graph and then moves backwards. So when a function has a return value moved to memory, that slot is marked occupied already on all call sites, so there will be no overlaps.